File: Java/BlazeServer/src/com/BlazeServer/Messaging/ProtoBuffEnvelope.java
package com.BlazeServer.Messaging;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.security.GeneralSecurityException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.KeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.DHPublicKeyParameters;
import org.bouncycastle.crypto.util.PublicKeyFactory;

import com.BlazeServer.Crypto.ServerAuthority;
import com.BlazeServer.Helper.EncryptionHelper;
import com.google.protobuf.Message;

public class ProtoBuffEnvelope extends MessageEnvelope {

	private final MessageRegistry registry;
	private boolean encryptionEnabled = false;
	private ServerAuthority sa = null;
	private AsymmetricKeyParameter remotePublicKey = null;
	
	public ProtoBuffEnvelope(MessageRegistry registry) {
		this.registry = registry;
	}
	
	@Override
	public void Serialize(Object messageObject, OutputStream stream) throws IOException {
		DataOutputStream dos = new DataOutputStream(stream);
		
		Message message = (Message)messageObject;

		if( encryptionEnabled ) {
			
			String privateKey = sa.generateAgreementValue(remotePublicKey).toString(16);
			byte [] buffer = message.toByteArray();
			try {
				
				// read utf8 string
				String encrypted = EncryptionHelper.encrypt(message.getDescriptorForType().getFullName(), privateKey); 
				//dos.writeBytes(encrypted + "\r\n");

				dos.writeUTF(encrypted);
				
				byte [] cipherbuffer = EncryptionHelper.encrypt(buffer, privateKey);
				dos.writeInt(cipherbuffer.length);
				dos.write(cipherbuffer, 0, cipherbuffer.length);
				
			} catch (InvalidKeyException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (NoSuchAlgorithmException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (NoSuchPaddingException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (InvalidAlgorithmParameterException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IllegalBlockSizeException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (BadPaddingException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			
			
		} else {
		
			dos.writeUTF(message.getDescriptorForType().getFullName());
			
			byte [] buffer = message.toByteArray();
			dos.writeInt(buffer.length);
			dos.write(buffer, 0, buffer.length);
		}

		dos.flush();
	}

	@Override
	public Object Deserialize(InputStream stream) throws IOException {
		
		DataInputStream dis = new DataInputStream(stream);
        String descriptorName = dis.readUTF();

        if( encryptionEnabled ) {
        	
			String privateKey = sa.generateAgreementValue(remotePublicKey).toString(16);
			try {
				descriptorName = EncryptionHelper.decrypt(descriptorName, privateKey);
		        System.out.println(descriptorName);		        

		        byte[] messageBuffer = new byte[dis.readInt()];
		        dis.read(messageBuffer);
				
		        byte[] decryptedBytes = EncryptionHelper.decrypt(messageBuffer, privateKey);
		        
		        Object value = null;
		        try {
					value = objectFromBuffer(descriptorName, decryptedBytes);
				} catch (Exception e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
		        
		        return value;
		        
			} catch (KeyException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (InvalidAlgorithmParameterException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IllegalBlockSizeException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (BadPaddingException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (GeneralSecurityException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
        	
        	return null;
        	
        } else {
	        System.out.println(descriptorName);
	       
	        byte[] messageBuffer = new byte[dis.readInt()];
	        dis.read(messageBuffer);
	        
	        Object value = null;
	        try {
				value = objectFromBuffer(descriptorName, messageBuffer);
			} catch (Exception e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
	        
	        return value;
        }
	}
	
	public void enableEncryption(ServerAuthority sa, byte [] publicKey) {
		this.sa = sa;
		
		try {
			remotePublicKey = new DHPublicKeyParameters(
					((DHPublicKeyParameters)PublicKeyFactory.createKey(publicKey)).getY(), 
					sa.getParameters());
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		this.encryptionEnabled = true;
	}
	
	private Object objectFromBuffer(String descriptorName, byte [] message) throws Exception {
		
		Class<?> clazz = registry.getTypeForDescriptor(descriptorName);

		if( clazz != null ) {
			Method parseFromMethod = clazz.getMethod("parseFrom", byte[].class);
			return parseFromMethod.invoke(null, message);
		}
		
		return null;
	}

}

File: SocketServer.Shared/Messaging/ProtoBuffEnvelope.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Google.ProtocolBuffers;
using Google.ProtocolBuffers.DescriptorProtos;
using Google.ProtocolBuffers.Descriptors;
using log4net;
using SocketServer.Shared.Reflection;
using SocketServer.Shared.Serialization;
using System.Reflection;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;
using SocketServer.Crypto;
using log4net.Core;

namespace SocketServer.Shared.Messaging
{
    public class ProtoBuffEnvelope : MessageEnvelope
    {
        private static ILog logger = LogManager.GetLogger(typeof(ProtoBuffEnvelope));

        private readonly MessageRegistry registry;

        private bool encryptionEnabled = false;
        private AsymmetricKeyParameter remotePublicKey = null;
        private ServerAuthority serverAuthority = null;
        
        public ProtoBuffEnvelope(MessageRegistry messageRegistry)
        {
            this.registry = messageRegistry;
        }

        public void DisableEncryption()
        {
            encryptionEnabled = false;
            this.serverAuthority = null;
            this.remotePublicKey = null;
        }

        public void EnableEncryption(ServerAuthority serverAuthority, byte[] encodedPublicKey)
        {
            encryptionEnabled = true;
            this.serverAuthority = serverAuthority;
            this.remotePublicKey = new DHPublicKeyParameters(
                ((DHPublicKeyParameters)PublicKeyFactory.CreateKey(encodedPublicKey)).Y, serverAuthority.Parameters);

        }

        public override void Serialize(object messageObject, Stream stream)
        {
            IMessage message = (IMessage)messageObject;
            string descriptorName = message.DescriptorForType.FullName;

            logger.Logger.Log(typeof(ProtoBuffEnvelope), Level.Finer, string.Format("sending {0}", descriptorName), null);

            if (encryptionEnabled)
            {
                string privateKey = serverAuthority.GenerateAgreementValue(remotePublicKey).ToString(16);

                RijndaelCrypto crypto = new RijndaelCrypto();
                string encryptedName = crypto.Encrypt(descriptorName, privateKey);

                stream.Write(encryptedName.SerializeUTF());

                //StringBuilder str = new StringBuilder();
                //str.AppendFormat("{0}\r\n", encryptedName);

                //byte[] nameBytes = MessageEncoding.GetBytes(str.ToString());
                //stream.Write(nameBytes, 0, nameBytes.Length);

                byte[] messageBytes = message.ToByteArray();
                byte[] encryptedBytes = crypto.Encrypt(messageBytes, privateKey);
                stream.Write(encryptedBytes.Serialize());
            }
            else
            {
                //StringBuilder str = new StringBuilder();
                //str.AppendFormat("{0}\r\n", descriptorName);

                stream.Write(descriptorName.SerializeUTF());

                //byte[] nameBytes = MessageEncoding.GetBytes(str.ToString());
                //stream.Write(nameBytes, 0, nameBytes.Length);

                byte[] messageBytes = message.ToByteArray();
                stream.Write(messageBytes.Serialize());
            }
        }

        public override object Deserialize(Stream stream)
        {
            if (encryptionEnabled)
            {
                string descriptorName = stream.ReadUTF();
                int length = stream.ReadInt();
                byte[] buffer = stream.Read(length);

                string privateKey = serverAuthority.GenerateAgreementValue(remotePublicKey).ToString(16);
                RijndaelCrypto crypto = new RijndaelCrypto();

                descriptorName = crypto.Decrypt(descriptorName, privateKey);

                logger.Logger.Log(typeof(ProtoBuffEnvelope), Level.Finer, string.Format("recieving {0}", descriptorName), null);

                Type t = registry.GetTypeForDescriptor(descriptorName);
                if (t != null)
                {
                    return ReflectionHelper.InvokeStaticMethodOnType(t, "ParseFrom", crypto.Decrypt(buffer, privateKey));
                }

            }
            else
            {
                string descriptorName = stream.ReadUTF();

                int length = stream.ReadInt();
                byte[] buffer = stream.Read(length);

                logger.Logger.Log(typeof(ProtoBuffEnvelope), Level.Finer, string.Format("recieving {0}", descriptorName), null);

                Type t = registry.GetTypeForDescriptor(descriptorName);
                if (t != null)
                {
                    return ReflectionHelper.InvokeStaticMethodOnType(t, "ParseFrom", buffer);
                }
            }

            return null;
        }
    }
}
